File: src/models/subscription.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, Text, ForeignKey, Numeric, Index, Boolean
from sqlalchemy.orm import relationship
from src.database import Base


class Subscription(Base):
    __tablename__ = "subscriptions"
    
    id = Column(Integer, primary_key=True, index=True)
    institution_id = Column(Integer, ForeignKey('institutions.id', ondelete='CASCADE'), nullable=False, index=True)
    plan_name = Column(String(100), nullable=False, index=True)
    status = Column(String(50), nullable=False, index=True)
    billing_cycle = Column(String(50), nullable=False)
    price = Column(Numeric(10, 2), nullable=False)
    currency = Column(String(3), nullable=False, default='INR')
    max_users = Column(Integer, nullable=True)
    max_storage_gb = Column(Integer, nullable=True)
    features = Column(Text, nullable=True)
    start_date = Column(DateTime, nullable=False)
    end_date = Column(DateTime, nullable=True)
    trial_end_date = Column(DateTime, nullable=True)
    grace_period_end = Column(DateTime, nullable=True)
    canceled_at = Column(DateTime, nullable=True)
    next_billing_date = Column(DateTime, nullable=True)
    auto_renew = Column(Boolean, default=True, nullable=False)
    external_subscription_id = Column(String(255), nullable=True, index=True)
    razorpay_subscription_id = Column(String(255), nullable=True, index=True)
    razorpay_customer_id = Column(String(255), nullable=True, index=True)
    metadata = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    institution = relationship("Institution", back_populates="subscriptions")
    payments = relationship("Payment", back_populates="subscription", cascade="all, delete-orphan")
    invoices = relationship("Invoice", back_populates="subscription", cascade="all, delete-orphan")
    usage_records = relationship("UsageRecord", back_populates="subscription", cascade="all, delete-orphan")
    
    __table_args__ = (
        Index('idx_subscription_institution_status', 'institution_id', 'status'),
        Index('idx_subscription_dates', 'start_date', 'end_date'),
        Index('idx_subscription_status', 'status'),
        Index('idx_subscription_next_billing', 'next_billing_date'),
        Index('idx_subscription_grace_period', 'grace_period_end'),
    )


class Payment(Base):
    __tablename__ = "payments"
    
    id = Column(Integer, primary_key=True, index=True)
    subscription_id = Column(Integer, ForeignKey('subscriptions.id', ondelete='CASCADE'), nullable=False, index=True)
    institution_id = Column(Integer, ForeignKey('institutions.id', ondelete='CASCADE'), nullable=False, index=True)
    amount = Column(Numeric(10, 2), nullable=False)
    currency = Column(String(3), nullable=False, default='INR')
    status = Column(String(50), nullable=False, index=True)
    payment_method = Column(String(50), nullable=True)
    razorpay_payment_id = Column(String(255), nullable=True, unique=True, index=True)
    razorpay_order_id = Column(String(255), nullable=True, index=True)
    razorpay_signature = Column(String(500), nullable=True)
    failure_reason = Column(Text, nullable=True)
    metadata = Column(Text, nullable=True)
    paid_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    subscription = relationship("Subscription", back_populates="payments")
    institution = relationship("Institution")
    
    __table_args__ = (
        Index('idx_payment_status', 'status'),
        Index('idx_payment_created', 'created_at'),
    )


class Invoice(Base):
    __tablename__ = "invoices"
    
    id = Column(Integer, primary_key=True, index=True)
    subscription_id = Column(Integer, ForeignKey('subscriptions.id', ondelete='CASCADE'), nullable=False, index=True)
    institution_id = Column(Integer, ForeignKey('institutions.id', ondelete='CASCADE'), nullable=False, index=True)
    invoice_number = Column(String(100), unique=True, nullable=False, index=True)
    status = Column(String(50), nullable=False, index=True)
    amount = Column(Numeric(10, 2), nullable=False)
    tax_amount = Column(Numeric(10, 2), nullable=False, default=0)
    total_amount = Column(Numeric(10, 2), nullable=False)
    currency = Column(String(3), nullable=False, default='INR')
    billing_period_start = Column(DateTime, nullable=False)
    billing_period_end = Column(DateTime, nullable=False)
    due_date = Column(DateTime, nullable=False)
    paid_at = Column(DateTime, nullable=True)
    invoice_url = Column(String(500), nullable=True)
    metadata = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    subscription = relationship("Subscription", back_populates="invoices")
    institution = relationship("Institution")
    
    __table_args__ = (
        Index('idx_invoice_status', 'status'),
        Index('idx_invoice_due_date', 'due_date'),
        Index('idx_invoice_created', 'created_at'),
    )


class UsageRecord(Base):
    __tablename__ = "usage_records"
    
    id = Column(Integer, primary_key=True, index=True)
    subscription_id = Column(Integer, ForeignKey('subscriptions.id', ondelete='CASCADE'), nullable=False, index=True)
    institution_id = Column(Integer, ForeignKey('institutions.id', ondelete='CASCADE'), nullable=False, index=True)
    metric_name = Column(String(100), nullable=False, index=True)
    metric_value = Column(Numeric(15, 2), nullable=False)
    recorded_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    period_start = Column(DateTime, nullable=False)
    period_end = Column(DateTime, nullable=False)
    metadata = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    
    subscription = relationship("Subscription", back_populates="usage_records")
    institution = relationship("Institution")
    
    __table_args__ = (
        Index('idx_usage_metric', 'metric_name'),
        Index('idx_usage_recorded', 'recorded_at'),
        Index('idx_usage_period', 'period_start', 'period_end'),
    )

File: frontend/src/api/subscription.ts
import axios from '@/lib/axios';

export interface Subscription {
  id: number;
  institution_id: number;
  plan_name: string;
  status: string;
  billing_cycle: string;
  price: number;
  currency: string;
  max_users: number | null;
  max_storage_gb: number | null;
  features: string | null;
  start_date: string;
  end_date: string | null;
  trial_end_date: string | null;
  grace_period_end: string | null;
  canceled_at: string | null;
  next_billing_date: string | null;
  auto_renew: boolean;
  external_subscription_id: string | null;
  razorpay_subscription_id: string | null;
  razorpay_customer_id: string | null;
  metadata: string | null;
  created_at: string;
  updated_at: string;
}

export interface Plan {
  name: string;
  display_name: string;
  description: string;
  monthly_price: number;
  quarterly_price: number;
  yearly_price: number;
  max_users: number | null;
  max_storage_gb: number | null;
  features: string[];
}

export interface Invoice {
  id: number;
  subscription_id: number;
  institution_id: number;
  invoice_number: string;
  status: string;
  amount: number;
  tax_amount: number;
  total_amount: number;
  currency: string;
  billing_period_start: string;
  billing_period_end: string;
  due_date: string;
  paid_at: string | null;
  invoice_url: string | null;
  metadata: string | null;
  created_at: string;
  updated_at: string;
}

export interface PaymentMethod {
  id: number;
  card_number: string;
  card_holder: string;
  expiry_month: string;
  expiry_year: string;
  is_default: boolean;
}

export interface AddOn {
  id: number;
  name: string;
  description: string;
  monthly_price: number;
  yearly_price: number | null;
  features: string[];
}

export interface UsageData {
  students_used: number;
  teachers_used: number;
  storage_used_gb: number;
}

export interface LimitsData {
  max_users: number | null;
  max_storage_gb: number | null;
}

export interface HistoryItem {
  id: number;
  type: 'created' | 'upgraded' | 'downgraded' | 'renewed' | 'canceled' | 'payment';
  title: string;
  description: string;
  date: string;
  amount?: number;
}

export interface SubscriptionData {
  subscription: Subscription;
  availablePlans: Plan[];
  invoices: Invoice[];
  paymentMethods: PaymentMethod[];
  addOns: AddOn[];
  activeAddOns: number[];
  usage: UsageData;
  limits: LimitsData;
  history: HistoryItem[];
}

const subscriptionApi = {
  async getInstitutionSubscription(): Promise<SubscriptionData> {
    const response = await axios.get('/api/v1/institution-admin/subscription');
    return response.data;
  },

  async getPlans(): Promise<Plan[]> {
    const response = await axios.get('/api/v1/subscriptions/plans');
    return response.data;
  },

  async upgradeSubscription(
    subscriptionId: number,
    data: { new_plan_name: string; billing_cycle: string }
  ): Promise<{ subscription: Subscription; prorated_amount: number; immediate_charge: boolean }> {
    const response = await axios.post(`/api/v1/subscriptions/${subscriptionId}/upgrade`, data);
    return response.data;
  },

  async downgradeSubscription(
    subscriptionId: number,
    data: { new_plan_name: string; billing_cycle: string }
  ): Promise<Subscription> {
    const response = await axios.post(`/api/v1/subscriptions/${subscriptionId}/downgrade`, data);
    return response.data;
  },

  async cancelSubscription(subscriptionId: number, immediate: boolean): Promise<Subscription> {
    const response = await axios.post(`/api/v1/subscriptions/${subscriptionId}/cancel`, {
      immediate,
    });
    return response.data;
  },

  async renewSubscription(subscriptionId: number): Promise<Subscription> {
    const response = await axios.post(`/api/v1/subscriptions/${subscriptionId}/renew`);
    return response.data;
  },

  async updateSubscription(
    subscriptionId: number,
    data: { auto_renew?: boolean; reminder_days?: number }
  ): Promise<Subscription> {
    const response = await axios.patch(`/api/v1/subscriptions/${subscriptionId}`, data);
    return response.data;
  },

  async getInvoices(subscriptionId?: number): Promise<Invoice[]> {
    const params = subscriptionId ? { subscription_id: subscriptionId } : {};
    const response = await axios.get('/api/v1/subscriptions/invoices/', { params });
    return response.data.invoices;
  },

  async downloadInvoice(invoiceId: number): Promise<Blob> {
    const response = await axios.get(`/api/v1/subscriptions/invoices/${invoiceId}/download`, {
      responseType: 'blob',
    });
    return response.data;
  },

  async addPaymentMethod(data: {
    card_number: string;
    card_holder: string;
    expiry_month: string;
    expiry_year: string;
    cvv: string;
  }): Promise<PaymentMethod> {
    const response = await axios.post('/api/v1/institution-admin/payment-methods', data);
    return response.data;
  },

  async deletePaymentMethod(methodId: number): Promise<void> {
    await axios.delete(`/api/v1/institution-admin/payment-methods/${methodId}`);
  },

  async setDefaultPaymentMethod(methodId: number): Promise<void> {
    await axios.post(`/api/v1/institution-admin/payment-methods/${methodId}/set-default`);
  },

  async enableAddOn(addOnId: number): Promise<void> {
    await axios.post(`/api/v1/institution-admin/add-ons/${addOnId}/enable`);
  },

  async disableAddOn(addOnId: number): Promise<void> {
    await axios.post(`/api/v1/institution-admin/add-ons/${addOnId}/disable`);
  },
};

export default subscriptionApi;
